### PR TITLE
[WIP] Distance return type

### PR DESCRIPTION
**Pull request by dalg24**
an attempt to avoid computing square roots

@hydroo the current version should improve the performance of kNN. ~I hadn't had a chance to tackle radius searches yet.~  same for radius search now :)

### CODE REVIEW

File: cmake/Dependencies.cmake [hunk 1]
```diff
@@ -1,9 +1,9 @@
 TRIBITS_PACKAGE_DEFINE_DEPENDENCIES(
   SUBPACKAGES_DIRS_CLASSIFICATIONS_OPTREQS
     Utils                 packages/Utils                      ST  REQUIRED
-    Interface             packages/Interface                  ST  REQUIRED
+    Interface             packages/Interface                  ST  OPTIONAL
     Search                packages/Search                     ST  REQUIRED
-    Discretization        packages/Discretization             ST  REQUIRED
-    Meshfree              packages/Meshfree                   ST  REQUIRED
+    Discretization        packages/Discretization             ST  OPTIONAL
+    Meshfree              packages/Meshfree                   ST  OPTIONAL
```
Comment (Rombur) [inline cmake/Dependencies.cmake]: This feels very arbitrary to have only `Search` and `Utils` as required packages. What can you do with the `Search` alone? I think that `Interface` and `Meshfree` should be required.

File: packages/Search/src/details/DTK_DetailsAlgorithms.hpp [hunk 2]
```diff
@@ -69,22 +70,87 @@ bool isValid( Sphere const &s )
            ( s.radius() >= 0. );
 }
 
+struct DistanceReturnType
+{
+  private:
+    double sq = 0.;
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsAlgorithms.hpp]: If the data is going to be private, use a `class` not a `struct`

File: packages/Search/src/details/DTK_DetailsAlgorithms.hpp [hunk 3]
```diff
@@ -101,10 +167,13 @@ double distance( Point const &point, Box const &box )
 
 // distance point-sphere
 KOKKOS_INLINE_FUNCTION
-double distance( Point const &point, Sphere const &sphere )
+DistanceReturnType distance( Point const &point, Sphere const &sphere )
 {
-    return KokkosHelpers::max(
+    // FIXME this triggers implicit conversion from DistanceReturnType to double
+    // which is precisely what I am trying to avoid...
+    double tmp = KokkosHelpers::max(
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsAlgorithms.hpp]: can you explicitly cast to `DistanceReturnType` to prevent the implicit cast?

**Comment by hydroo** [comment]
Added the benchmarks to #393 [(link to comment)](https://github.com/ORNL-CEES/DataTransferKit/pull/393#issuecomment-389223706)

The linked excel sheet contains the raw data.

File: cmake/Dependencies.cmake (hunk 1 above)
Comment (dalg24) [reply cmake/Dependencies.cmake]: It takes for ever to build on my laptop and I cannot disable if it is required. Anyway I am not planning on committing these changes to maste, this is a work in progress.

**Comment by dalg24** [comment]
@hydroo so we mainly see the benefits with CUDA. Did you trade the cout in the implicit conversion to double operator for a Kokkos::abort like we discussed?

File: packages/Search/src/details/DTK_DetailsAlgorithms.hpp (hunk 3 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsAlgorithms.hpp]: What triggers the implicit conversion to double here is the operator- (the subtraction).  I am not sure how I want to deal with that one yet.  This is not call anywhere in the traversals selected for the benchmark.

Comment (sslattery) [reply packages/Search/src/details/DTK_DetailsAlgorithms.hpp]: Do we implement a templated `operator -` to deal with this?

**Comment by hydroo** [comment]
@dalg24 I did. In 16dd6bea i won't, OK?

**Comment by hydroo** [comment]
Summitdev, GCC 5.4.0, 100M values, 1M queries

What      | master | KNN 8d7053f  | Radius 16dd6be
--------- | ------ | ------------ | --------------
Serial    | 1      |  .955        |  .877
OpenMP 1  | 1      |  .969        |  .882
OpenMP 2  | 1      |  .969        |  .885
OpenMP 4  | 1      |  .972        |  .886
OpenMP 8  | 1      |  .974        |  .894
OpenMP 16 | 1      |  .975        |  .921 
CUDA      | 1      |  .884        | 1.092

(knn at 16dd6be looks the same as 8d7053f)

Since Distance makes up ~40% of serial CPU I'd expect more of an improvement.
(I'll have a look with the profiler soon)

+ @dalg24 why would CUDA be slower? Any hunch?

File: packages/Search/src/details/DTK_DetailsAlgorithms.hpp (hunk 3 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsAlgorithms.hpp]: > Do we implement a templated operator - to deal with this?

What do you mean?

**Comment by hydroo** [comment]
@dalg24 this didn't change performance at all over 16dd6be  (within 1%).

**Comment by hydroo** [comment]
I can't comment on code at the moment.
Maybe rename the class to just `Distance` or perhaps even better `SquaredDouble`.
Or said differently: I don't see why you would put `ReturnType` in its name.

**Comment by dalg24** [comment]
Rebased onto master and measured performance again.

device \ version  | master (924aec6) | head (fcf7957)
---| --- | ---
**cuda (P100)**
construction | 0.237284 | 0.22116 |
kNN          | 0.404008 | 0.384168 |
radius       | 0.270756 | 0.304295 |
**openmp (16 threads)**
construction | 0.319492 | 0.319183 |
kNN          | 0.356629 | 0.344704 |
radius       | 0.379063 | 0.316911 |
**serial**
construction | 3.19788 | 3.16436 |
kNN          | 4.18104 | 4.04033 |
radius       | 4.52308 | 3.72549 |

with `--values=10000000 --queries=1000000`

**Comment by dalg24** [comment]
For the record, I checked that no conversion happens when we run the example bvh drivers by doing:
```diff
diff --git a/packages/Search/src/details/DTK_DetailsAlgorithms.hpp b/packages/Search/src/details/DTK_DetailsAlgorithms.hpp
index 22186f0..91d91c7 100644
--- a/packages/Search/src/details/DTK_DetailsAlgorithms.hpp
+++ b/packages/Search/src/details/DTK_DetailsAlgorithms.hpp
@@ -89,7 +89,11 @@ struct DistanceReturnType
                            "requires non-negative floating-point value" );
     }
     // TODO consider removing implicit conversion to double
-    KOKKOS_INLINE_FUNCTION operator double() const { return std::sqrt( sq ); }
+    KOKKOS_INLINE_FUNCTION operator double() const
+    {
+        Kokkos::abort( "oh no" );
+        return std::sqrt( sq );
+    }
     KOKKOS_INLINE_FUNCTION bool operator<( DistanceReturnType const &rhs ) const
     {
         return sq < rhs.sq;
```
I am not sure how/whether we could test for this.

File: packages/Utils/test/tstKokkosHelpers.cpp [hunk 4]
```diff
@@ -9,11 +9,26 @@
...
+// cppreference.com
+
+TEUCHOS_UNIT_TEST( KokkosHelpers, is_nan )
+{
+    using DataTransferKit::KokkosHelpers::isNan;
+    TEST_ASSERT( isNan( NAN ) );
+    TEST_ASSERT( !isNan( INFINITY ) );
+    TEST_ASSERT( !isNan( 0.0 ) );
+    TEST_ASSERT( !isNan( DBL_MIN / 2.0 ) );
+    TEST_ASSERT( isNan( 0.0 / 0.0 ) );
```
Comment (dalg24) [inline packages/Utils/test/tstKokkosHelpers.cpp]: Clang Sanitizer didn't like this dividing by zero :/

**Comment by aprokop** [comment]
> Rebased onto master and measured performance again.

So, in summary, it drastically improves `radius` for OpenMP, while slows down CUDA?

**Comment by dalg24** [comment]
> drastically improves radius for OpenMP, while slows down CUDA?

Improves radius and knn search everywhere except on cuda where radius is slower which I can't explain at the moment

**Comment by aprokop** [comment]
Moved to arborx/arborx#72

**Pull request closed without merging**